Support for Minecraft 1.11 and 1.12

Bjarno opened the pull request:
I've added support for blocks added in Minecraft 1.11 and 1.12.

The only thing that still needs to be done are coloured beds (see question later). I also found a bug with how pistons are rendered which I will try to solve in another pull request.

I support all rotations of observers and shulker boxes (1.11), and all the new blocks (and their colours/rotation combinations) from 1.12, except for beds as mentioned before.

Edit: Beds have been implemented, and two bugs with pistons have been resolved. See below for more details.

Bjarno commented:
Some extra explanation:

I've used the Shulker entity files to generate the textures, as in the blocks textures folder, there is only a texture for the top one. As the side textures consist of two parts, I merge those.

Concrete blocks and concrete powder blocks are regular blocks without any special properties.

The Glazed Terracotta block contains a direction (in the NBT tag this is stored in "facing", but also in the data value). It's quite a lot of manual code to define how each block should look like, and could possible be improved (code quality wise), but the result should be "perfect" as far as I can see.

If needed, I can provide a screenshot of a test world, and both topdown and isometric renders.

Bjarno commented:
**Question:** What is the cleanest way to read NBT data? When I looked over the source, I did not find anything. The changes I did all work with the damage/data value for orientation/colour. But as beds both have a orientation and a "side" (as before) and a colour... The colour, according to the Minecraft Wiki (http://minecraft.gamepedia.com/Bed) is stored in an NBT tag.

Bjarno commented:
I found a way to read the colour of beds, I probably need to change the renderer so it can also use some information from the block/tile entities list, or find a way to circumvent it, but reading it is not a problem anymore. I've got quite a busy week coming up, but I expect it to be working latest this friday.

el0hh commented:
Thank you and great Job!

m0r13 commented:
Thanks a lot for your contribution! I'll have a look at your commits and maybe add some notes.

Reading NBT data is done in the chunk class, if you need an example for that. Unfortunately there is no use of NBT entity data when rendering blocks yet, but that should be possible to implement as all the block data is read from NBT too.

m0r13 commented on src/mapcraftercore/renderer/blockimages.cpp:
```diff
@@ -109,6 +109,59 @@ bool DoubleChestTextures::load(const std::string& filename, int texture_size) {
+            load_single(base_filename + "black.png", 15, texture_size));
+}
+
+bool ShulkerTextures::load_single(const std::string& filename, int color_index, int texture_size) {
```
Please use the function name style of the remaining code base: `loadSingle`.

m0r13 commented on src/mapcraftercore/renderer/blockimages.cpp:
```diff
@@ -109,6 +109,59 @@ bool DoubleChestTextures::load(const std::string& filename, int texture_size) {
+        return false;
+    }
+    // if the image is 64px wide, the shulker box images are 16x16
+    int ratio = image.getHeight() / 64;
```
I'm not sure about the possible shulker box image sizes, but maybe use a float as ratio?

Bjarno replied:
I "copied" this from `ChestTextures::load()` (and others). In these implementations, an integer is also used used to calculate the ratio between the actual texture, and the reference texture dimensions.

Normally the ratio should be a power of 2, thus should fit in an int. Also, as these are working on bitmaps, you can't copy half pixels, so I would have to do an explicit conversion later anyway.

m0r13 replied:
Alright then :)

Bjarno replied:
I think it still shows up as "changes requested", due to this not being an issue. I don't think I can flag it like that however 😛

m0r13 replied:
Should be okay now.

m0r13 commented on src/mapcraftercore/renderer/blockimages.h:
```diff
@@ -205,9 +231,10 @@ class TextureResources {
-	bool loadChests(const std::string& normal_png, const std::string& normal_double_png,
-			const std::string& ender_png, const std::string& trapped_png,
-			const std::string& trapped_double_png);
+    bool loadChests( const std::string& normal_png,
```
Please use tabs as indendation.

m0r13 commented on src/mapcraftercore/renderer/blockimages.h:
```diff
@@ -230,6 +257,7 @@ class TextureResources {
 
 	ChestTextures normal_chest, ender_chest, trapped_chest;
 	DoubleChestTextures normal_double_chest, trapped_double_chest;
+    ShulkerTextures shulker_textures;
```
Indendation.

m0r13 commented on src/mapcraftercore/renderer/blocktextures.cpp:
```diff
@@ -197,6 +229,22 @@ BlockTextures::BlockTextures()
 	  GLASS_SILVER("glass_silver"),
 	  GLASS_WHITE("glass_white"),
 	  GLASS_YELLOW("glass_yellow"),
+      GLAZED_TERRACOTTA_BLACK("glazed_terracotta_black"),
```
I see the indendation here has changed too. Have you added these texture images manually? There is a make target to generate this code if you added the texture images to the `src/data/textures`-directory: Just run `make gen_texture_code` in case you didn't do this.

Bjarno replied:
I actually did not know, that explains the indentation errors here. Normally my IDE should detect these automagically, but apparently that failed. I did not knew that `make gen_texture_code` would update the source files, I will update this.

m0r13 commented on src/mapcraftercore/renderer/renderviews/isometric/blockimages.cpp:
```diff
@@ -729,6 +772,35 @@ void IsometricBlockImages::createDispenserDropper(uint16_t id, const RGBAImage&
+
+    setBlockImage(id, 1, down_image);
+
+    // createBlock(id, 1, side.rotate(ROTATE_270), top.rotate(ROTATE_180));
```
Dead code left.

m0r13 commented on src/mapcraftercore/renderer/renderviews/isometric/blockimages.cpp:
```diff
@@ -836,6 +908,7 @@ void IsometricBlockImages::createPiston(uint16_t id, bool sticky) { //  id 29, 3
 	createBlock(id, 0, side.rotate(ROTATE_180), back);
 	createBlock(id, 1, side, front);
 
+    // TODO: Bug fix rotations
```
Indendation.

m0r13 commented on src/mapcraftercore/renderer/renderviews/isometric/blockimages.cpp:
```diff
@@ -996,6 +1069,53 @@ void IsometricBlockImages::createStairs(uint16_t id, const RGBAImage& texture) {
 }
 
+void IsometricBlockImages::createShulkerBox(uint16_t id, int color_index, const ShulkerTextures& textures) { // id 219 - 234
+    int offset = color_index * ShulkerTextures::DATA_SIZE;
```
Indendation.

m0r13 commented on src/mapcraftercore/renderer/renderviews/isometric/blockimages.cpp:
```diff
@@ -2338,6 +2458,76 @@ void IsometricBlockImages::createBlocks() {
 	createBlock(216, 4, t.BONE_BLOCK_TOP, t.BONE_BLOCK_SIDE, t.BONE_BLOCK_SIDE); // east-west
 	createBlock(216, 8, t.BONE_BLOCK_SIDE, t.BONE_BLOCK_TOP, t.BONE_BLOCK_SIDE); // north-south
 	// --
+    createObserver(218); // observer
```
Indendation.

m0r13 commented on src/mapcraftercore/renderer/renderviews/isometric/blockimages.h:
```diff
@@ -257,7 +257,9 @@ class IsometricBlockImages : public AbstractBlockImages {
-	void createRotatedBlock(uint16_t id, uint16_t extra_data, const RGBAImage& front_texture,
+	void createGlazedTerracotta(uint16_t id, const RGBAImage& texture);
+	void createRotatedBlock(uint16_t id, uint16_t extra_data, const RGBAImage& texture);
+    void createRotatedBlock(uint16_t id, uint16_t extra_data, const RGBAImage& front_texture,
```
Indendation.

m0r13 commented on src/mapcraftercore/renderer/renderviews/topdown/blockimages.cpp:
```diff
@@ -1631,6 +1638,73 @@ void TopdownBlockImages::createBlocks() {
+	createGlazedTerracotta(249, t.GLAZED_TERRACOTTA_RED); // red
+	createGlazedTerracotta(250, t.GLAZED_TERRACOTTA_BLACK); // black
+    // concrete --
+    setBlockImage(251, 0, t.CONCRETE_WHITE); // white
```
Indendation.

m0r13 commented:
If you want to implement rendering of beds too and not merge the implemented features yet, you can also continue working on that feature for this PR and we just keep the PR open.

Bjarno commented:
Except for testing and cleanup, beds are almost implemented on the beds branch in my fork. I expect to merge with master later today.

Bjarno commented:
Okay, so I implemented beds.

Quick overview:

- While parsing a chunk, information about bed colours is saved in `std::unordered_map<int, EntityBed> entities_beds`, which is a mapping from an integer (unique key for the local position) to a EntityBed struct, which contains the colour
- The colour can be looked up by the renderer, using `Chunk::getExtraBlockData(const LocalBlockPos& pos)`
- When looking for the texture to draw while rendering a block/tile, this extra_data is passed along (all non-bed blocks have extra data 0)
- For assigning bed textures, a seperate map is created (`std::unordered_map<uint32_t, RGBAImage> block_images_bed`) which contains a mapping from data+extra_data to the actual texture (we only store beds in here, so we don't need the id anymore)
- If `getBlock()` sees that a bed is given, it finds the correct texture, based on the data and the extra_data (and not just id and data), this texture is then used to render

It might be possible to optimize this, but it should be a good starting point.

---

I also fixed two bugs with the rendering of pistons. In the topdown view, the wrong texture was used (sticky texture for non-sticky pistons and vice versa). And in isometric view, the north and east faces were drawn incorrectly.

Bjarno commented:
I'm probably going to do a bit of refactoring later this day, so don't accept/merge it yet. Might result in a small speed up, which is always a good thing.

Bjarno commented:
Okay, I did some cleanup and optimisation steps. If you read the commit description of https://github.com/mapcrafter/mapcrafter/pull/256/commits/0d2ba30cbb751c9842e24a8ae83358e88937eac2 it should be clear what changes I did to make it more future-proof and possibly faster.

Unless some changes are requested, I think this should be the last change needed to have adequate support for 1.11 and 1.12 blocks.

m0r13 commented on src/mapcraftercore/renderer/renderviews/isometric/tilerenderer.cpp:
```diff
@@ -194,17 +194,16 @@ void IsometricTileRenderer::renderTile(const TilePos& tile_pos, RGBAImage& tile)
-			uint16_t id = current_chunk->getBlockID(local);
+			// now get block id and block data
+			uint16_t id, data, extra_data;
+			current_chunk->getBlockInfo(local, id, data, extra_data);
```
Performance-wise I thought it would be good to fetch the block ID first, and then the data if it's not air, but maybe it's not that relevant. Have you noticed any performance changes?

Bjarno replied:
For the small test I used, not really, but I expect that it might have more influence on larger worlds (given that most worlds container more air than any other "block").

It can be avoided if the id is retrieved, as before... but as `getBlockExtraData` needs the id, it needs to be passed as an argument to it, which is not really good code quality in my opinion. Hiding it behind `Chunk::getBlockInfo` seemed a decent solution.

But as performance for these kind of programs is important, I'll revert back to the previous implementation, and will pass the id to `Chunk::getBlockExtraData`. Or `Chunk::getBlockExtraData` first needs to retrieve the id, which is probably going to be even slower (twice the same access for the id), so let's not do that.

m0r13 replied:
Calling `Chunk::getBlockExtraData` from the rendering code manually with the id sounds fine.

m0r13 commented:
Code looks fine now. Last quick thing I would be interested in... Can you maybe show me a screenshot of the new blocks rendered? Then I don't have to setup a new world and all that stuff. :)

Bjarno commented:
In-game screenshots:
![2017-06-18_15 38 49](https://user-images.githubusercontent.com/6963292/27261084-551c38f0-543c-11e7-829a-65900e0f59ec.png)
![2017-06-18_15 39 02](https://user-images.githubusercontent.com/6963292/27261085-56eaa0d6-543c-11e7-8547-d3195979b9e5.png)
![2017-06-18_15 39 11](https://user-images.githubusercontent.com/6963292/27261086-580308aa-543c-11e7-8257-075100c91252.png)

Mapcrafter renders:
![image](https://user-images.githubusercontent.com/6963292/27261136-efc27e64-543c-11e7-972f-5ded30356bb4.png)
![image](https://user-images.githubusercontent.com/6963292/27261140-0293a82e-543d-11e7-90ba-dfacc343937f.png)
![image](https://user-images.githubusercontent.com/6963292/27261124-daf93f36-543c-11e7-8682-637bd3ae2d1a.png)

World download: 
[flatmap.zip](https://github.com/mapcrafter/mapcrafter/files/1083113/flatmap.zip)

m0r13 commented:
Looks great! Thanks a lot for your contribution! I'm going to merge it now.
